### PR TITLE
feat: UnitedDeployment (OpenKruise) surge support

### PR DESCRIPTION
**Pull request by NamanBhat**
## Summary

Adds OpenKruise **UnitedDeployment** support to the eviction-autoscaler surge path. When PDB-protected pods belong to a UnitedDeployment, the controller now surges the UnitedDeployment's own `spec.replicas` (routing the delta to a chosen subset) instead of the child Deployment ? which the UD controller would otherwise revert.

## What changed

- **Target resolution** (`pdb_to_evictionautoscaler_controller.go`): `discoverDeployment` now walks pod ? ReplicaSet ? Deployment ? **UnitedDeployment** owner refs and returns the resolved kind. EvictionAutoScaler CR + PDB owner-ref are set to the UD when applicable.
- **UnitedDeploymentWrapper** (`ud_wrapper.go`) + **UnitedDeploymentSurgeApplier** (`ud_surge_applier.go`): surge via UD spec, snapshot original per-subset topology for exact revert, idempotent.
- **Subset selection**: `surge-subset` annotation override > drain-displaced subset (pods on cordoned nodes) > dominant subset fallback.
- Kruise scheme registration (`cmd/main.go`), RBAC markers + `role.yaml`, README section.

## Testing

- 7 UD unit tests (subset selection, snapshot/restore, idempotency, maxSurge).
- Full existing controller suite passes (no regressions to Deployment/HPA/KEDA paths).
- Live-validated on an AKS cluster with a spot/on-demand UnitedDeployment.

## Follow-ups (out of scope)

- Orphan protection (finalizer + stale-window backstop) ? kept consistent with existing appliers, none of which have it today.
- envtest coverage with a vendored kruise CRD.

## Prerequisite

OpenKruise must be installed (UnitedDeployment CRD present).

### CODE REVIEW

File: internal/controller/pdb_to_evictionautoscaler_controller.go [hunk 1]
```diff
@@ -279,7 +287,7 @@ func (r *PDBToEvictionAutoScalerReconciler) discoverDeployment(ctx context.Conte
 				replicaSet := &appsv1.ReplicaSet{}
 				err = r.Get(ctx, k8s_types.NamespacedName{Name: ownerRef.Name, Namespace: pdb.Namespace}, replicaSet)
 				if apierrors.IsNotFound(err) {
-					return "", "", fmt.Errorf("error fetching ReplicaSet: %v", err)
+					return "", "", "", fmt.Errorf("error fetching ReplicaSet: %v", err)
 				}
```
[the remark this answers is not in the document]
Comment (NamanBhat) [reply internal/controller/pdb_to_evictionautoscaler_controller.go:294]: Thanks. This `r.Get`/`IsNotFound` handling is pre-existing on the plain Deployment owner-walk (unchanged by this PR) and applies to all target kinds, not just UnitedDeployment. To keep this PR focused on UD support I've left it as-is and will track the error-handling hardening (return on non-NotFound errors; try the next pod/ownerRef on a NotFound ReplicaSet) as a separate follow-up. Note the new `unitedDeploymentOwner` helper this PR adds does propagate non-NotFound errors correctly.

File: internal/controller/evictionautoscaler_controller.go [hunk 2]
```diff
@@ -203,6 +205,18 @@ func (r *EvictionAutoScalerReconciler) Reconcile(ctx context.Context, req ctrl.R
...
+		// For a UnitedDeployment target, default the surge to land on the subset
+		// the drain is displacing (overridable via the surge-subset annotation).
+		if udWrapper, isUD := target.(*UnitedDeploymentWrapper); isUD {
+			subset, subsetErr := displacedSubsetOnCordoned(ctx, r.Client, pdb)
+			if subsetErr != nil {
+				logger.Error(subsetErr, "failed to determine displaced UnitedDeployment subset; falling back to dominant subset")
+			} else if subset != "" {
+				logger.Info("Routing UnitedDeployment surge to displaced subset", "pdb", pdb.Name, "subset", subset)
+				udWrapper.SetPreferredSurgeSubset(subset)
+			}
```
[the remark this answers is not in the document]
Comment (NamanBhat) [reply internal/controller/evictionautoscaler_controller.go]: Fixed. Merged the two passes into a single `computeCordonedDrainStats` that returns both the displaced count and the dominant displaced subset from one pod+node listing; `countPodsOnCordoned` now delegates to it and the reconcile calls the combined helper once. (dd08bcf)

File: internal/controller/ud_surge_applier.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	target Surger
+}
+
+var _ SurgeApplier = &UnitedDeploymentSurgeApplier{}
+
+func (u *UnitedDeploymentSurgeApplier) ApplySurge(ctx context.Context, surgeReplicas int32) error {
+	u.target.SetReplicas(surgeReplicas)
+	u.target.AddAnnotation(EvictionSurgeReplicasAnnotationKey, strconv.FormatInt(int64(surgeReplicas), 10))
+	return u.client.Update(ctx, u.target.Obj())
+}
```
[the remark this answers is not in the document]
Comment (NamanBhat) [reply internal/controller/ud_surge_applier.go:40]: Added. `ud_surge_applier_test.go` now covers the ApplySurge -> RevertSurge round-trip with a fake client (asserting spec.replicas, snapshot + surge-marker annotations, and their removal on revert), plus a lost-snapshot revert that forces scale-down to the floor. Also added discoverDeployment UD-redirect tests (flag on/off) and wrapper edge-case tests. (dd08bcf)

**Comment by NamanBhat** [comment]
Pushed dd08bcf addressing the review feedback. Summary:

**Correctness / robustness**
- **Status-convergence guard:** surge now refuses to snapshot until `status.subsetReplicas` matches `spec.replicas`, so a subset that is transiently lagging in status is never pinned to a stale/zero count and scaled down.
- **nil `spec.replicas`:** `GetReplicas` now derives the total from the sum of `status.subsetReplicas` instead of defaulting to 1 (which over-provisioned the surge subset).
- **Snapshot safety:** `PrepareSurge` refuses to re-capture a snapshot while a surge is already active (prevents re-baselining off the surged state); `RestoreOriginal` deep-copies before its early return; `RevertSurge` scales the UD back down to the floor when the snapshot is lost so it can never be left stuck surged; the snapshot marshal error is no longer swallowed.
- **Single-pass cordoned-pod stats** (no more double pod/node listing).

**Config & observability**
- **Feature flag `ENABLE_UNITEDDEPLOYMENT_SURGE`** (default `false`), strictly validated in `main.go`, gating both the target redirect and the surge strategy ? consistent with the existing env-flag pattern.
- **Metric** `eviction_autoscaler_uniteddeployment_surge_failures_total{namespace,reason}` for the handled failure paths.

**Tests**
- Applier ApplySurge/RevertSurge round-trip + lost-snapshot scale-down (fake client), `discoverDeployment` UD-redirect on/off, and wrapper edge cases (nil replicas, not-converged, lost-snapshot guard, invalid-annotation fallback, ForceScaleDown). Full suite green.

**Deferred as follow-ups (to keep this PR focused):**
- Moving the topology snapshot from a UD annotation to the EvictionAutoScaler CR `.status` (the re-capture guard covers the practical risk today).
- envtest coverage with a vendored OpenKruise CRD (the `kruise-api` module ships types only).
- Orphan protection (finalizer + stale-window backstop) ? no surge applier has this today, so kept consistent.
- Hardening the pre-existing ReplicaSet `Get`/`IsNotFound` handling on the Deployment owner-walk.

**Comment by skuchipudi295** [comment]
### Code review

No issues found. Checked for bugs and CLAUDE.md compliance.
